pkg/dsl/dslmodule.go

```diff
@@ -0,0 +1,153 @@
+	"reflect"
+)
+
+type dslModuleImpl struct {
+	moduleID          t.ModuleID
+	eventHandlers     map[reflect.Type][]func(ev *eventpb.Event) error
+	conditionHandlers []func() error
+	outputEvents      *events.EventList
+	// contextStore is used to store and recover context on asynchronous operations such as signature verification.
+	contextStore cs.ContextStore[any]
+	// eventCleanupContextIDs is used to dispose the
+	eventCleanupContextIDs map[ContextID]struct{}
+}
+
+type Handle struct {
+	impl *dslModuleImpl
+}
+
+type ContextID = cs.ItemID
+
+// Module allows creating passive modules in a very natural declarative way.
+type Module interface {
+	modules.PassiveModule
+
+	// GetDslHandle returns an object that
+	GetDslHandle() Handle
+
+	// GetModuleID returns the identifier of the module.
+	// TODO: consider moving this method to modules.Module.
+	GetModuleID() t.ModuleID
+}
+
+// NewModule creates a new dsl module with a given id.
+func NewModule(moduleID t.ModuleID) Module {
+	return &dslModuleImpl{
+		moduleID:      moduleID,
+		eventHandlers: make(map[reflect.Type][]func(ev *eventpb.Event) error),
+		outputEvents:  &events.EventList{},
+		contextStore:  cs.NewSequentialContextStore[any](),
+	}
+}
+
+func (m *dslModuleImpl) GetDslHandle() Handle {
+	return Handle{m}
+}
+
+func (m *dslModuleImpl) GetModuleID() t.ModuleID {
+	return m.moduleID
+}
+
+// RegisterEventHandler registers an event handler for module m.
+// This event handler will be called every time an event of type EvTp is received.
+// TODO: consider adding a protoc plugin that would augment EvTp with a function Unwrap(), which would return the
+// 		 unwrapped event. Then it will be possible to pass the unwrapped event to the handler.
+func RegisterEventHandler[EvTp events.EventType](m Module, handler func(ev *EvTp) error) {
+	evTpPtrType := reflect.PointerTo(reflectutil.TypeOf[EvTp]())
+
+	m.GetDslHandle().impl.eventHandlers[evTpPtrType] = append(m.GetDslHandle().impl.eventHandlers[evTpPtrType],
+		func(event *eventpb.Event) error {
+			evTpPtr := ((any)(event.Type)).(*EvTp)
+			return handler(evTpPtr)
+		})
+}
+
+// UponCondition registers a special type of handler that will be invoked each time after processing a batch of events.
+// The handler is assumed to represent a conditional action: it is supposed to check some predicate on the state
+// and perform actions if the predicate evaluates is satisfied.
+func UponCondition(m Module, handler func() error) {
+	impl := m.GetDslHandle().impl
+	impl.conditionHandlers = append(impl.conditionHandlers, handler)
+}
+
+// StoreContext stores the given data and returns an automatically deterministically generated unique id.
+// The data can be later recovered or disposed of using this id.
+func (h Handle) StoreContext(context any) ContextID {
+	return h.impl.contextStore.Store(context)
+}
+
+// CleanupContext schedules a disposal of context with the given id after the current batch of events is processed.
```

Comment:
What is the reason (I'm sure there is one) for not just disposing of the context right away?

Reply:
One of the central design choices is that each event can be processed by multiple handlers.
So, you cannot dispose of the context after it is processed by just one handler. You could do it after the whole event is processed, but I decided that the semantics would be clearer if we do it after the whole batch is processed.

Reply:
Makes sense.

Reply:
I added on that.